juju/juju
# Implement secret-rotate hook

This PR wires up the previously added secret rotation worker to trigger the execution of a new `secret-rotate` hook whenever a secret should be rotated. Whenever a secret is successfully rotated, the unit agent  will, during hook commit, make a new `SecretsRotated` API call to record the last rotation time. This allows the secrets rotation worker to re-trigger a rotate event if the unit agent crashes, whether or not leadership changes before the agent restarts (only the unit leader runs the rotate hook). Secret rotation takes precedence over running other hooks like config-changed etc.

So the main components are:
- SecretsRotated api 
- new secrets hook type
- new secrets hook resolver

There's a couple of other drive by fixes:
- wire up feature flags to k8s unit agents
- adjust the valid secret URL syntax to allow secret paths with "-"
- tweak GetSecretValue API params to accept either an ID or URL

## QA steps

boostrap and deploy a charm with a simple `secret-rotated` hook, eg
```
#!/bin/bash
juju-log "secret-rotate $JUJU_SECRET_URL"
exit 0
```
create a secret without any rotate interval
`juju exec --unit ubuntu/0 "secret-create password hello=world"`

`juju show-status-log ubuntu/0` will show no secret hook runs

update the rotate interval
`juju exec --unit ubuntu/0 "secret-update password --rotate=1m"`

now observe the `secret-rotate` hook running every minute
```
$ juju show-status-log ubuntu/0
Time                        Type       Status       Message
...
17 Sep 2021 07:47:04+10:00  juju-unit  idle         
17 Sep 2021 07:48:05+10:00  juju-unit  executing    running secret-rotate hook for secret://app/ubuntu/password
17 Sep 2021 07:48:05+10:00  juju-unit  idle         
17 Sep 2021 07:49:06+10:00  juju-unit  executing    running secret-rotate hook for secret://app/ubuntu/password
17 Sep 2021 07:49:06+10:00  juju-unit  idle 
```

juju debug log shows the hook was run with the correct secret url in the context
```
...
unit-ubuntu-0: 07:48:05 INFO unit.ubuntu/0.juju-log secret-rotate secret://app/ubuntu/password
...
```

disable rotate interval and secret-rotate hook no longer runs
`juju exec --unit ubuntu/0 "secret-update password --rotate=0"`
